### DISCUSSION TITLE
결제실패 에러 관련 문의드립니다.

### DISCUSSION THREAD

**Discussion by billyLee1004** [commerce-api-naver/commerce-solutions#81] [묻고 답하기]
안녕하세요 에이아이닷엠 빌리 입니다.

결제 실패 에러 관련 문의드립니다. 
error-msg: N/A 

으로 와서왜 결제가 실패 했는지 궁금합니다 

결제실패 로그입니다. 
createdDate: '2024-12-07T09:03:14.071+0000',
eventDate: '2024-12-07T09:03:14.276+0000',
changeType: 'PAYMENT_FAIL',
plan: { id: 'PLN_5Bil0jv88SEScQB1Rz4Y9V', grade: 'DEFAULT', name: '스몰' },
subscriptionStatus: 'ON_PAYMENT_FAIL',
round: 2,
transactionId: 'CSP241207100305B1T001S11832',
transactionType: 'PAYMENT',
totalAmount: 26900,

그리고 저희가 의심하는 건 저번에 문의 드렸던 비노출 상품 planid로 결제 요청햇을 때 실패하는 게 아닌가 싶은데 
노출 되었을 때 구독하고 그 상품이 비노출 되어도 그 상품으로 계속 결제가 되는게 맞을까요?

**Comment by commerce-api-naver** [comment]
안녕하세요. 커머스API 이민석입니다.

아시다시피 노출된 요금제가 향후 솔루션 사정에 따라 비노출되더라도 이전에 (대상 요금제로) 구독을 진행하신 구독자 분들께는 비노출 상태인 구독제 정보가 그대로 유지됩니다.
이때 `plainId` 값 등의 데이터 뿐만 아니라 구독 당시 정한 정기결제 조건 역시 그대로 유지되며, 요금제의 노출 여부와는 관계 없이 정기결제는 정상적으로 진행됩니다.

알려주신 이벤트 훅 데이터에 해당하는 솔루션 구독 정기결제 실패건은 구독자 분의 결제 수단 문제로 정기결제가 실패하는 것을 보입니다.
구독자 분께서 등록하신 결제 수단으로 결제가 불가능한 상황이기에 구독자 분을 통해 스마트스토어센터 네이버페이 비즈월렛의 충전 잔고를 확인해보시거나, 솔루션 구독 당시 비즈월렛 정기결제를 신청하신 경우 정하신 결제 수단으로 정상적인 결제가 가능한지 확인해보도록 가이드해보시기 바랍니다.

**Reply by billyLee1004** [reply]
답변 감사합니다. 
그리고 추가적으로
1. error-msg: N/A 이렇게 왔는데 더 자세한 에러 로그를 볼 수 있는 방법이 있을까요?  
2. 구독자에게 네이버에서 저희가 따로 연락할 방법이 있을까요? 그리고 직접 연락해서 가이드 하라는 얘기인거죠?
3. 결제 실패를 계속해서 강제해지 되었을 때 구독자의 솔루션 목록에서 해지목록으로 이동되나요? 
4. 구독 실패시 구독자분이 구독 결제가 실패 했다는 알림이 스마트 스토어 센터에 표시 될까요?

**Reply by commerce-api-naver** [reply]
@billyLee1004 님, 커머스API 이민석입니다.

대상하는 내용은 커머스솔루션마켓이 아닌 솔루션 시스템에서 처리하는 메시지로 추정됩니다.
알려주신 이벤트 훅 데이터 일부 내용으로 커머스솔루션마켓에서 '셀롸이트' 솔루션 시스템으로 전달된 이벤트 훅 발송 기록을 검토해보았으나 `error-msg` 필드 혹은 "N/A" 형식의 값으로 된 데이터는 전달되지 않은 것으로 보입니다.

커머스솔루션의 [이벤트 훅 JSON 스킴](https://apicenter.commerce.naver.com/ko/basic/solution-doc#section/%EA%B0%9C%EB%B0%9C%ED%95%98%EA%B8%B0/%EC%9D%B4%EB%B2%A4%ED%8A%B8-%ED%9B%85-%EC%97%B0%EB%8F%99-%EC%9A%94%EC%86%8C-%EA%B0%80%EC%9D%B4%EB%93%9C)에서도 `error-msg`라는 이름의 필드는 지원하지 않습니다.
혹시 솔루션 시스템에서 자체적으로 처리하는 부분이 있는지 확인해보시기 바랍니다.

특정 시간대 발송된 이벤트 훅 전문의 확인이 필요하신 경우 요청주시면 민감 정보를 제외하고 이 쓰레드로 공유드리겠습니다.

**Reply by commerce-api-naver** [reply]
@billyLee1004 님, 커머스API 이민석입니다.

문의글 수정으로 추가문의 주신 건을 아래와 같이 답변드립니다.

### 두 번째 문의 내용에 답변
> 구독자에게 네이버에서 저희가 따로 연락할 방법이 있을까요? 그리고 직접 연락해서 가이드 하라는 얘기인거죠?

#### 1. 커머스API센터에서는 솔루션 개발사가 개별 구독자에게 연락할 수 있는 수단을 지원하지 않습니다.

#### 2. 개별 연락 가이드는 솔루션 개발/운용사 재량입니다.
뒤이어 답변드리겠으나 정기 결제 실패는 구독자 분께 개별적으로 안내됩니다.
솔루션 개발/운용사 판단에 따라 공지성 안내가 필요하신 경우를 대비하여 답변드린 내용으로 이해해주시면 될 것 같습니다.

### 세 번째 문의 내용에 답변
> 결제 실패를 계속해서 강제해지 되었을 때 구독자의 솔루션 목록에서 해지목록으로 이동되나요?

네, 맞습니다.

### 네 번째 문의 내용에 답변
> 구독 실패시 구독자분이 구독 결제가 실패 했다는 알림이 스마트 스토어 센터에 표시 될까요?

네, 맞습니다.
솔루션 정기구독 결제 실패의 경우 스마트스토어센터 알림과 구매자 분에 대한 이메일 안내 등이 발송됩니다.
- https://help.sell.smartstore.naver.com/faq/content.help?faqId=13782

**Reply by billyLee1004** [reply]
답변 감사합니다

**Reply by billyLee1004** [reply]
안녕하세요 아래 결제 실패 시 이벤트 로그인데 
    eventId: '3ScOdfMLkKl1BFJRGdDCjN_1',
    createdDate: '2024-12-11T01:04:26.358+0000',
    eventDate: '2024-12-11T01:04:26.699+0000',
    changeType: 'PAYMENT_FAIL',
    plan: { id: 'PLN_5YQlfAUkU5Q3v6BuJX5AA9', grade: 'DEFAULT', name: '스몰' },
    subscriptionStatus: 'ON_PAYMENT_FAIL',
    round: 2,
    transactionId: 'CSP241211100316B1T001S11292',
    transactionType: 'PAYMENT',
    totalAmount: 29590,
    billingPlan: { id: 'PLN_5YQlfAUkU5Q3v6BuJX5AA9', grade: 'DEFAULT', name: '스몰' }
    
    PAYMENT_FAIL 이 왜 났는지 알 수 있을까요? 예를 들어 잔고부족인지,, 카드 유효기간이 지났는지..등등

**Reply by commerce-api-naver** [reply]
@billyLee1004 님 안녕하세요. 커머스API 이민석입니다.

해당 조건은 자칫 구독자 분의 프라이버시 문제와도 연결될 수 있는 문제라 공개 공간에서 답변을 드리기 어렵습니다.
비록 알려주신 이벤트 훅 데이터에서는 구독자를 특정할 수 있는 정보는 없으나 결제 조건과 관련된 내용은 개인의 프라이버시와 연결된 문제이기도 하기에 바로 확인이 어려움에 양해 부탁드립니다.

- 이벤트 훅의 `changeType`으로 "PAYMENT_FAIL" 값이 전달되는 경우는 대부분 기술적 원인은 아닙니다.
    - 기술적 원인에 의한 결제 실패(예: 커머스솔루션마켓 및 네이버페이 시스템 오류 등)는 일정 시간 후 자동 재결제를 시도할 수 있으나 통상적인 운용 상황에서는 발생 빈도가 극도로 낮습니다.
    - 대부분의 경우 구독자 분의 결제 조건에 따른 결제 실패 현상으로 이해해주시면 좋을 것 같습니다. (잔고 부족 등)
- 결제 실패에 대한 상세 내용은 구독자 분께 개별적으로 안내됩니다.
    - 앞선 답변과 같이 스마트스토어센터 알림, 개별 이메일 알림 등 구독자 분께 솔루션 정기결제 실패에 대한 상세 원인과 해결 방법을 개별적으로 안내해드리고 있습니다. 따라서 솔루션 운영 차원에서 결제 실패건을 매 번 구독자 분께 알려드리지 않으셔도 됩니다.

만일 솔루션 운용 과정에서 사용자 CS 처리 등 피치못할 사정으로 대상건의 상세한 결제 실패 사유 확인이 필요하실 경우 보유하고 계신 솔루션의 테스트 스토어용 계정으로 [스마트스토어 고객센터](https://help.sell.smartstore.naver.com/mail/form.help?alias=sssolution2)를 통해 문의해주시기 바랍니다.
- 상담분류: [커머스솔루션 - 솔루션 구독문제/개발이슈] 으로 선택 (필수)
- 상담내용 내 포함 권장:
    - 기술지원 포럼으로 문의 시 '구독자 프라이버시' 문제로 고객센터 문의로 답변받았음을 포함
    - 이 쓰레드의 URL을 문의 내용에 포함하는 것을 권장 (`https://github.com/commerce-api-naver/commerce-solutions/discussions/81#discussioncomment-11527551`)

**Reply by billyLee1004** [reply]
답변 감사합니다.
